src/index.ts

```diff
@@ -0,0 +1,272 @@
+import { pipe } from 'it-pipe'
+import { Pushable, pushableV } from 'it-pushable'
+import debug from 'debug'
+import { abortableSource } from 'abortable-iterator'
+import { encode } from './encode.js'
+import { decode } from './decode.js'
+import { restrictSize } from './restrict-size.js'
+import { MessageTypes, MessageTypeNames, Message } from './message-types.js'
+import { createStream } from './stream.js'
+import { toString as uint8ArrayToString } from 'uint8arrays'
+import type { AbortOptions } from '@libp2p/interfaces'
+import type { Duplex, Sink } from 'it-stream-types'
+import type { MuxedStream, Muxer, MuxedTimeline } from '@libp2p/interfaces/stream-muxer'
+
+const log = Object.assign(debug('libp2p:mplex'), {
+  error: debug('libp2p:mplex:error')
+})
+
+function printMessage (msg: Message) {
+  const output: any = {
+    ...msg,
+    type: `${MessageTypeNames[msg.type]} (${msg.type})`
+  }
+
+  if (msg.type === MessageTypes.NEW_STREAM) {
+    output.data = uint8ArrayToString(msg.data instanceof Uint8Array ? msg.data : msg.data.slice())
+  }
+
+  if (msg.type === MessageTypes.MESSAGE_INITIATOR || msg.type === MessageTypes.MESSAGE_RECEIVER) {
+    output.data = uint8ArrayToString(msg.data instanceof Uint8Array ? msg.data : msg.data.slice(), 'base16')
+  }
+
+  return output
+}
+
+export interface MplexStream extends Duplex<Uint8Array> {
+  source: Pushable<Uint8Array>
+  close: () => void
+  abort: (err?: Error) => void
+  reset: () => void
+  timeline: MuxedTimeline
+  id: string
+}
+
+export interface MplexOptions extends AbortOptions {
+  onStream?: (...args: any[]) => void
+  onStreamEnd?: (...args: any[]) => void
+  maxMsgSize?: number
+}
+
+export class Mplex implements Muxer {
+  static multicodec = '/mplex/6.7.0'
+
+  public sink: Sink<Uint8Array>
+  public source: AsyncIterable<Uint8Array>
+
+  private _streamId: number
+  private readonly _streams: { initiators: Map<number, MplexStream>, receivers: Map<number, MplexStream> }
+  private readonly _options: MplexOptions
+  private readonly _source: { push: (val: Message) => void, end: (err?: Error) => void }
+
+  /**
+   * @class
+   * @param {object} options
+   * @param {function(*)} options.onStream - Called whenever an inbound stream is created
+   * @param {function(*)} options.onStreamEnd - Called whenever a stream ends
+   * @param {AbortSignal} options.signal - An AbortController signal
+   */
+  constructor (options?: MplexOptions) {
+    options = options ?? {}
+    options = typeof options === 'function' ? { onStream: options } : options
+
+    this._streamId = 0
+    this._streams = {
+      /**
+       * Stream to ids map
+       */
+      initiators: new Map<number, any>(),
```

Comment:
is value actually MuxedStream?

Reply:
Yes, well, a `Stream` rather than a `MuxedStream` now - the idea being the muxing should be transparent to the called.  I've updated the types for the map to reflect this.